### ISSUE TITLE
Flat R_Curve from val.py

### ISSUE THREAD

**Issue by mrken823** [ultralytics/yolov5#5432]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Validation

### Bug

I'm using Google Colab and sadly it interrupts due to daily limit. It was ok at first since I can always resume training and get the useful plots from val.py.

However my R_Curve is simply a horizontal line (same recall for all confidence level), while P_curve works normally. Is this a bug?

![R_curve](https://user-images.githubusercontent.com/42903058/139678619-c6200ba6-b3f1-4274-8dfd-5ae8dbc5939e.png)
![P_curve](https://user-images.githubusercontent.com/42903058/139678629-b7472e13-96ce-44e2-8c2c-1b6158fee575.png)
![F1_curve](https://user-images.githubusercontent.com/42903058/139678635-d89f2ee4-5231-4bd4-82fe-bc88415c76b2.png)
![PR_curve](https://user-images.githubusercontent.com/42903058/139678641-88b782b0-6b54-4fe9-a2b2-a203300525b7.png)
![confusion_matrix](https://user-images.githubusercontent.com/42903058/139678645-5c1e9dff-a4e5-4849-99b2-a0e0da5b73dd.png)

### Environment

K80 on Google Colab

### Minimal Reproducible Example

%%time
%cd "/G/My Drive/Folder M/Folder B/yolov5/yolov5/"
!python train.py --resume --img 1280 --batch 4 --epochs 64 --data './GPR data.yaml' --cfg './yolov5l6.yaml' --weights './runs/train/GPR_yolov5l6/weights/last.pt' --name GPR_yolov5l6  --cache --adam

!python val.py --weights './runs/train/GPR_yolov5l6/weights/best.pt' --data './GPR data.yaml' --batch-size 4 --img 1280 --save-hybrid --name GPR_yolov5l6_val --half --task train

!python val.py --weights './runs/train/GPR_yolov5l6/weights/best.pt' --data './GPR data.yaml' --batch-size 4 --img 1280 --save-hybrid --name GPR_yolov5l6_val --half --task val

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Zengyf-CVer** [comment]
@mrken823 
According to recall=TP/(TP+FN), it can be seen from your Recall graph that recall is always equal to 1, so FN=0 can be deduced, indicating that there are no FN samples in your training set. Please double check your data set allocation problem and instance allocation problem.

**Comment by glenn-jocher** [comment]
@mrken823 P and R are plotted at max F1, which is 1.00 confidence in your case. So R should be plotted at 1.0 conf rather than 0.0, not sure why that's not the case. Can you submit a reproducible example?

**Comment by mrken823** [comment]
Updated Minimal Reproducible Example.

Currently it's a single class object detection.
outputs from train and val.py seems normal (?)

![image](https://user-images.githubusercontent.com/42903058/139787583-16509dfe-03a1-4c2d-9e5b-545717a2d0fc.png)

![image](https://user-images.githubusercontent.com/42903058/140001833-8825bc06-ed86-453b-b137-7ed095dbf3ee.png)

Additionally, I encounter another problem generating plots after finishing training, seems like some problem with validation after training ends. When I run val.py with --task train, similar horizontal R_curve is generated
![image](https://user-images.githubusercontent.com/42903058/139787833-dd883a0e-5a15-40ac-85b5-b03d06a245b5.png)

**Comment by glenn-jocher** [comment]
@mrken823 I tried to reproduce this on a single class dataset, but everything seems fine for me. Recall legend is correct, it is showing the max Recall value and indicating the x axis position for that value (0.0).

```bash
!python train.py --img 640 --batch 16 --epochs 1 --data GlobalWheat2020.yaml --weights yolov5s.pt --cache

train: weights=yolov5s.pt, cfg=, data=GlobalWheat2020.yaml, hyp=data/hyps/hyp.scratch.yaml, epochs=1, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, evolve=None, bucket=, cache=ram, image_weights=False, device=, multi_scale=False, single_cls=False, adam=False, sync_bn=False, workers=8, project=runs/train, name=exp, exist_ok=False, quad=False, linear_lr=False, label_smoothing=0.0, patience=100, freeze=0, save_period=-1, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
github: up to date with https://github.com/ultralytics/yolov5 ✅
YOLOv5 🚀 v6.0-43-g19c8760 torch 1.9.0+cu111 CUDA:0 (Tesla P100-PCIE-16GB, 16280.875MB)

hyperparameters: lr0=0.01, lrf=0.1, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
Weights & Biases: run 'pip install wandb' to automatically track and visualize YOLOv5 🚀 runs (RECOMMENDED)
TensorBoard: Start with 'tensorboard --logdir runs/train', view at http://localhost:6006/
Overriding model.yaml nc=80 with nc=1

                 from  n    params  module                                  arguments                     
  0                -1  1      3520  models.common.Conv                      [3, 32, 6, 2, 2]              
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     18816  models.common.C3                        [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  2    115712  models.common.C3                        [128, 128, 2]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  3    625152  models.common.C3                        [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1   1182720  models.common.C3                        [512, 512, 1]                 
  9                -1  1    656896  models.common.SPPF                      [512, 512, 5]                 
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]          
 24      [17, 20, 23]  1     16182  models.yolo.Detect                      [1, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model Summary: 270 layers, 7022326 parameters, 7022326 gradients, 15.8 GFLOPs

Transferred 343/349 items from yolov5s.pt
Scaled weight_decay = 0.0005
optimizer: SGD with parameter groups 57 weight, 60 weight (no decay), 60 bias
albumentations: version 1.0.3 required by YOLOv5, but version 0.1.12 is currently installed
train: Scanning '../datasets/GlobalWheat2020/labels/arvalis_1.cache' images and labels... 3376 found, 46 missing, 0 empty, 0 corrupted: 100% 3422/3422 [00:00<?, ?it/s]
train: Caching images (4.2GB ram): 100% 3422/3422 [00:48<00:00, 71.04it/s]
val: Scanning '../datasets/GlobalWheat2020/labels/ethz_1' images and labels...747 found, 0 missing, 0 empty, 0 corrupted: 100% 747/747 [00:00<00:00, 1130.52it/s]
val: New cache created: ../datasets/GlobalWheat2020/labels/ethz_1.cache
val: Caching images (0.9GB ram): 100% 747/747 [00:08<00:00, 86.55it/s]
Plotting labels... 

autoanchor: Analyzing anchors... anchors/target = 5.94, Best Possible Recall (BPR) = 1.0000
Image sizes 640 train, 640 val
Using 4 dataloader workers
Logging results to runs/train/exp4
Starting training for 1 epochs...

     Epoch   gpu_mem       box       obj       cls    labels  img_size
       0/0     3.48G   0.08911    0.2727         0       841       640: 100% 214/214 [01:22<00:00,  2.60it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100% 24/24 [00:10<00:00,  2.37it/s]
                 all        747      49603      0.469      0.647      0.514      0.134

1 epochs completed in 0.026 hours.
Optimizer stripped from runs/train/exp4/weights/last.pt, 14.5MB
Optimizer stripped from runs/train/exp4/weights/best.pt, 14.5MB

Validating runs/train/exp4/weights/best.pt...
Fusing layers... 
Model Summary: 213 layers, 7012822 parameters, 0 gradients, 15.8 GFLOPs
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100% 24/24 [01:04<00:00,  2.70s/it]
                 all        747      49603      0.469      0.645      0.513      0.134
Results saved to runs/train/exp4
```
<img width="500" src="https://user-images.githubusercontent.com/26833433/139843623-c74c3dae-6736-49a1-ab6c-bf73e25cd11b.png">
<img width="500" src="https://user-images.githubusercontent.com/26833433/139843627-e819128b-cfba-4705-aab2-7484f3715d7f.png">
<img width="500" src="https://user-images.githubusercontent.com/26833433/139843629-9d90c526-9d78-43d0-a09b-67e777b235c8.png">
<img width="500" src="https://user-images.githubusercontent.com/26833433/139843633-791d2625-293e-4aaf-8556-530aa6e3bb9c.png">
<img width="500" src="https://user-images.githubusercontent.com/26833433/139843616-21bd760f-06d9-4318-bd36-bf996a0dde77.png">

**Comment by mrken823** [comment]
May I know the output of
1. val.py --task train
2. val.py --task val?

Could that a problem in my dataset that interrupted validation after last epoch which led to plots not generated in ./runs/train/xxx/*.png?
and that same problem also led to the flat R_curve issue when I try to run val.py afterwards?

In my case, I see the "Wall time..." printed so I guess it's not Colab's problem since the cell did finish running train.py, even if the validation suddenly stopped after 1 iteration. However, only results.csv is left and no plots are generated even if all 64 epochs are finished
.
![image](https://user-images.githubusercontent.com/42903058/140022665-63c5f351-34a4-4467-a785-12d251a5d369.png)

**Comment by glenn-jocher** [comment]
@mrken823 val.py --task argument uses either your train, val, or test split. See argparser for details:
https://github.com/ultralytics/yolov5/blob/042f02ff9b095e402a937b24c6da6d6e83e554b2/val.py#L308

**Comment by mrken823** [comment]
> @mrken823 val.py --task argument uses either your train, val, or test split. See argparser for details:
> 
> https://github.com/ultralytics/yolov5/blob/042f02ff9b095e402a937b24c6da6d6e83e554b2/val.py#L308

I asked because both outputs from val.py did not generated plots like a success/complete train.py run did on the final epoch. Recall-confidence curve is still weird from val.py and no plots are generated from train.py recently :(

**Comment by glenn-jocher** [comment]
@mrken823 if you believe plots are missing from train or val commands upon completion please submit a bug report with a reproducible example so we can get started debugging. Thank you!

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃
